subsys/secure_storage/src/aead/secure_backend_aead.c

```diff
@@ -0,0 +1,338 @@
+{
+	psa_storage_create_flags_t data_flags;
+	size_t data_size;
+	psa_status_t status;
+
+	if (p_info == NULL) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	/* Get size & flags */
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_FLAGS,
+				    (void *)&data_flags, sizeof(data_flags));
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_SIZE,
+				    (void *)&data_size, sizeof(data_size));
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	p_info->capacity = SECURE_STORAGE_MAX_ASSET_SIZE;
+	p_info->size = data_size;
+	p_info->flags = data_flags;
+
+	return PSA_SUCCESS;
+}
+
+psa_status_t secure_get(const psa_storage_uid_t uid, const char *prefix, size_t data_offset,
+			size_t data_length, void *p_data, size_t *p_data_length)
+{
+	psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+	psa_storage_create_flags_t data_flags;
+	struct aead_additional_data additional_data;
+	uint8_t key_buf[AEAD_KEY_SIZE + 1];
+	uint8_t nonce[AEAD_NONCE_SIZE];
+	size_t object_data_size;
+	size_t aead_out_size;
+	size_t data_size;
+
+	if (data_length == 0 || p_data == NULL || p_data_length == NULL) {
+		return PSA_ERROR_INVALID_ARGUMENT;
+	}
+
+	if ((data_offset + data_length) > SECURE_STORAGE_MAX_ASSET_SIZE) {
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	/* Get flags then size */
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_FLAGS,
+				    (void *)&data_flags, sizeof(data_flags));
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_SIZE,
+				    (void *)&data_size, sizeof(data_size));
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	/* Calculate the exact output size of encrypted buffer */
+	object_data_size = secure_storage_aead_get_encrypted_size(data_size);
+
+	if (object_data_size > AEAD_MAX_BUF_SIZE) {
+		return PSA_ERROR_NOT_SUPPORTED;
+	}
+
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_NONCE,
+				    nonce, sizeof(nonce));
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	status = storage_get_object(uid, prefix, SECURE_STORAGE_AEAD_BACKEND_FILENAME_SUFFIX_DATA,
+				    aead_buf, object_data_size);
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	status = secure_storage_get_key(uid, key_buf, AEAD_KEY_SIZE);
+	if (status != PSA_SUCCESS) {
+		return status;
+	}
+
+	additional_data.uid = uid;
+	additional_data.flags = data_flags;
+	additional_data.size = data_length;
+
+	status = secure_storage_aead_decrypt(key_buf, AEAD_KEY_SIZE, nonce, AEAD_NONCE_SIZE,
+					     (void *)&additional_data, sizeof(additional_data),
+					     aead_buf, object_data_size, data_buf,
+					     SECURE_STORAGE_MAX_ASSET_SIZE, &aead_out_size);
+
+	/* Clean up */
+	safe_memset(key_buf, sizeof(key_buf), 0, AEAD_KEY_SIZE);
```

Comment:
idk. there are many returns from this functions that will not result in this cleanup.

Reply:
True moved it to one clean up block at the end